controllers/provisioning_controller.go

```diff
@@ -114,82 +113,72 @@ func (r *ProvisioningReconciler) readProvisioningCR(ctx context.Context) (*metal
 	return instance, nil
 }
 
-// Reconcile updates the cluster settings when the Provisioning
+// sync updates the cluster settings when the Provisioning
 // resource changes
-func (r *ProvisioningReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	// provisioning.metal3.io is a singleton
-	// Note: this check is here to make sure that the early startup configuration
-	// is correct. For day 2 operatations the webhook will validate this.
-	if req.Name != metal3iov1alpha1.ProvisioningSingletonName {
-		klog.Info("ignoring invalid CR", "name", req.Name)
-		return ctrl.Result{}, nil
-	}
-
+func (r *ProvisioningController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
 	// Make sure ClusterOperator exists
 	err := r.ensureClusterOperator(nil)
 	if err != nil {
-		return ctrl.Result{}, err
-	}
-
-	result := ctrl.Result{}
-	if !r.WebHookEnabled {
-		if provisioning.WebhookDependenciesReady(r.OSClient) {
-			klog.Info("restarting to enable the webhook")
-			os.Exit(1)
-		}
-		// Keep checking for our webhook dependencies to be ready, so we can
-		// enable the webhook.
-		result.RequeueAfter = 5 * time.Minute
+		return err
 	}
 
 	enabled, err := r.isEnabled()
 	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "could not determine whether to run")
+		return errors.Wrap(err, "could not determine whether to run")
 	}
 	if !enabled {
 		// set ClusterOperator status to disabled=true, available=true
 		// We're disabled; don't requeue
-		return ctrl.Result{}, errors.Wrapf(
+		return errors.Wrapf(
 			r.updateCOStatus(ReasonUnsupported, "Nothing to do on this Platform", ""),
 			"unable to put %q ClusterOperator in Disabled state", clusterOperatorName)
 	}
 
 	baremetalConfig, err := r.readProvisioningCR(ctx)
 	if err != nil {
 		// Error reading the object - requeue the request.
-		return ctrl.Result{}, err
+		return err
 	}
 	if baremetalConfig == nil {
 		// Provisioning configuration not available at this time.
 		// Cannot proceed wtih metal3 deployment.
 		klog.Info("Provisioning CR not found")
-		return result, nil
+		return nil
+	}
+
+	if !r.webHookEnabled {
+		if provisioning.WebhookDependenciesReady(r.osClient) &&
+			slice.Contains(baremetalConfig.ObjectMeta.Finalizers, metal3iov1alpha1.ProvisioningFinalizer) {
+			klog.Info("restarting to enable the webhook")
+			os.Exit(1)
```

Comment:
this reads oddly.  Your controller is restarting?

Reply:
We can't enable webhooks from the start, as service-ca is not ready to inject the ca in time. So we start without the webhook (which is fine as the installer creates the provisioning CR - so it's correct). Then later when service-ca is up, we restart with the webhook enabled. It is definitely one of the ugly things I'd like to fix with library-go.